gcore/gdal_priv.h

```diff
@@ -1812,6 +1813,9 @@ class CPL_DLL GDALDriver : public GDALMajorObject
         CSLConstList papszVectorTranslateArguments,
         GDALProgressFunc pfnProgress, void *pProgressData) = nullptr;
 
+    /* Returns a (possibly null) pointer to the Subdataset informational function */
+    GDALSubdatasetInfo *(*pfnGetSubdatasetInfoFunc)() = nullptr;
```

Comment:
I would have make pfnGetSubdatasetInfoFunc take a const char* argument with the subdataset string. That way this saves the GDALGetSubdatasetInfo::IsSubdatasetSyntax() method, and passing the subdataset name as an argument of each method of GDALGetSubdatasetInfo
(even if that method was kept, I don't think it should be exposed in the C API, as it is mostly an implementation detail of GDALSubdatasetInfo)